p2p/crypto/secio/rw.go

```diff
@@ -105,66 +101,27 @@ func (r *etmReader) NextMsgLen() (int, error) {
 	return r.msg.NextMsgLen()
 }
 
-func (r *etmReader) drainBuf(buf []byte) int {
-	if r.buf == nil {
-		return 0
-	}
-
-	n := copy(buf, r.buf)
-	r.buf = r.buf[n:]
-	return n
-}
-
 func (r *etmReader) Read(buf []byte) (int, error) {
 	r.Lock()
 	defer r.Unlock()
 
-	// first, check if we have anything in the buffer
-	copied := r.drainBuf(buf)
-	buf = buf[copied:]
```

Comment:
no i believe this is right as is -- (this seems to be way more complicated than it should be, though).

- `buf` here is the user passed in `buf` which will receive the data
- `r.drainBuf(buf)` will copy into `buf` all the stuff in `r.buf` that fits into `buf`.
- it also updates `r.buf` to be `r.buf[n:]` (i.e. already copied `buf[0:n]` into `buf`)
- `copied` is the number of bytes copied into `buf`. this may be a number smaller than `len(buf)`, and if so, we want to set `buf` to be `buf[copied:]` so that further copies into buf only fill up the _remaining_ (unused) bytes.
- below, we copy more into `buf`, and in that case we only write into the original `buf[copied:]`, because that's what hasn't been written to.

all of this is useless anyway because exit if `copied > 0`, which seems to have been added later, as a short circuit for simplicity.